apps/next/src/pages/Approve/components/hardware/keystone-usb/components/Pending.tsx

```diff
@@ -0,0 +1,92 @@
+import { FC, useEffect, useRef } from 'react';
+import { Stack, Typography } from '@avalabs/k2-alpine';
+import { Action, ActionStatus } from '@core/types';
+import { DisplayData } from '@avalabs/vm-module-types';
+import { StateComponentProps } from '../types';
+import { PendingKeystoneCircles } from './PendingKeystoneCircles';
+import { useTranslation } from 'react-i18next';
+import { useKeystoneUsbContext } from '@core/ui';
+import { createKeystoneTransport } from '@keystonehq/hw-transport-webusb';
+import { resolve } from '@core/common';
+
+type PendingProps = StateComponentProps & {
+  action: Action<DisplayData>;
+};
+
+export const Pending: FC<PendingProps> = ({ state, approve, action }) => {
+  const { t } = useTranslation();
+  const { hasKeystoneTransport } = useKeystoneUsbContext();
+  const hasApprovedRef = useRef(false);
+
+  useEffect(() => {
+    // Reset flag when state is not pending or action is not pending
+    if (state !== 'pending' || action.status !== ActionStatus.PENDING) {
+      hasApprovedRef.current = false;
+      return;
+    }
+
+    // Only call approve when device is ready and we haven't called it yet
+    if (hasKeystoneTransport && !hasApprovedRef.current) {
+      // Verify device is actually ready by testing transport creation
+      // This ensures the device can handle signing requests
+      const verifyAndApprove = async () => {
+        try {
+          // Make sure that the transport is available and the device is ready
+          const [usbTransport] = await resolve(createKeystoneTransport());
+          if (!usbTransport) {
+            throw new Error('Transport not available');
+          }
+
+          // Device is ready, now call approve
+          hasApprovedRef.current = true;
+          await approve();
+        } catch (error) {
+          console.error('Approval error:', error);
+        } finally {
+          hasApprovedRef.current = false;
+        }
+      };
+
+      // Small delay to ensure state is stable
+      const timeoutId = setTimeout(() => {
+        // Check again before calling (device might have disconnected during delay)
+        if (
+          state === 'pending' &&
+          action.status === ActionStatus.PENDING &&
+          hasKeystoneTransport &&
+          !hasApprovedRef.current
+        ) {
+          verifyAndApprove();
+        } else {
+          hasApprovedRef.current = false;
+        }
+      }, 500);
```

Comment:
I feel like all these timeouts are going to come back and bite us someday. Why is it not possible to wait for proper ("stable") state without them?

We can try to hop on a call to find a better solution if you'd like.